File: Android/app/src/main/java/live/ditto/pos/core/data/repository/LocationsRepository.kt
package live.ditto.pos.core.data.repository

import com.ditto.kotlin.Ditto
import com.ditto.kotlin.DittoSyncSubscription
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import live.ditto.ditto_wrapper.DittoManager
import live.ditto.pos.core.data.demo.LocationSeed
import live.ditto.pos.core.data.locations.Location
import live.ditto.pos.core.data.observeAsFlow
import live.ditto.pos.settings.AppSettings
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Owns the `locations` collection and the active-location state.
 * `setActiveLocation` is the only writer — it persists to [AppSettings] and
 * applies [DittoManager]'s sync group. Other per-collection repositories
 * subscribe to `locationIdFlow()` and re-register their own subscriptions
 * when it emits. Mirrors iOS `LocationsRepository`.
 */
@Singleton
class LocationsRepository @Inject constructor(
    private val dittoManager: DittoManager,
    private val appSettings: AppSettings
) {
    private val ditto: Ditto get() = dittoManager.requireDitto()
    private var subscription: DittoSyncSubscription? = null
    private val repoScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /**
     * One shared observer on `locations`, mirroring iOS's single `allLocations`
     * publisher. `by lazy` so `requireDitto()` isn't touched at injection time,
     * and `stateIn` so every consumer replays the latest list off one observer
     * instead of opening its own.
     */
    private val allLocations by lazy {
        ditto.store
            .observeAsFlow<Location>("SELECT * FROM ${Location.COLLECTION_NAME}")
            .map { locations -> locations.filter { it.id in LocationSeed.demoLocationIds } }
            .stateIn(repoScope, SharingStarted.WhileSubscribed(STOP_TIMEOUT_MILLIS), emptyList())
    }

    init {
        // Restore the persisted active location on first injection (mirrors
        // iOS) — but only if it's still one of the seven demo locations. A
        // stale id (e.g. a legacy custom location that no longer exists) is
        // cleared so the user is forced to re-pick, and the sync group is reset
        // to default. The sync group applies as part of `setActiveLocation`.
        repoScope.launch {
            val saved = appSettings.locationId()
            if (saved.isNotEmpty()) {
                if (saved in LocationSeed.demoLocationIds) {
                    setActiveLocation(saved)
                } else {
                    setActiveLocation("")
                }
            }
        }
    }

    fun startSubscription() {
        if (subscription != null) return
        subscription = try {
            ditto.sync.registerSubscription(
                "SELECT * FROM ${Location.COLLECTION_NAME}"
            )
        } catch (error: Throwable) {
            reportSubscriptionFailure("subscribe locations", error)
        }
    }

    /**
     * Switch the active location. Persists, applies the sync group, and downstream
     * repositories react via `locationIdFlow()`. Pass an empty string to clear
     * the selection and reset the sync group back to default.
     */
    suspend fun setActiveLocation(locationId: String) {
        appSettings.setLocationId(locationId = locationId)
        if (locationId.isNotEmpty()) {
            dittoManager.setSyncGroup(locationId)
        } else {
            dittoManager.resetSyncGroup()
        }
    }

    /** Reactive stream of the active location id. Empty string when not set. */
    fun locationIdFlow(): Flow<String> = appSettings.locationIdFlow()

    fun observeAllLocations(): Flow<List<Location>> = allLocations

    /**
     * The active location resolved against the synced `locations` collection.
     * Null when nothing is selected, or when the saved id isn't one of the demo
     * locations — [observeAllLocations] already filters to those, so no separate
     * validity check is needed. Mirrors iOS `LocationsRepository.currentLocation`.
     */
    fun currentLocationFlow(): Flow<Location?> =
        combine(locationIdFlow(), observeAllLocations()) { locationId, locations ->
            if (locationId.isEmpty()) null else locations.firstOrNull { it.id == locationId }
        }

    private companion object {
        const val STOP_TIMEOUT_MILLIS = 5_000L
    }
}

File: iOS/DittoPOS/Data/Repositories/LocationsRepository.swift
//
//  LocationsRepository.swift
//  DittoPOS
//
//  Copyright © 2026 DittoLive Incorporated. All rights reserved.
//

import Combine
import DittoSwift

/// Owns the `locations` collection and the in-memory state for the currently
/// selected location. `allLocations` is filtered to the seven demo location
/// IDs defensively. `currentLocationId` is the single source of truth for
/// "which location is active" — UI calls `setActive(_:)` to switch and the
/// per-collection repositories react via their own subscriptions to
/// `$currentLocationId`. `currentLocation` is a convenience derivation for
/// views.
@MainActor final class LocationsRepository: ObservableObject {

    @Published private(set) var allLocations: [Location] = []
    @Published private(set) var currentLocationId: String?
    @Published private(set) var currentLocation: Location?

    private let dittoManager: DittoManager
    private var ditto: Ditto { dittoManager.ditto }
    private var store: DittoStore { ditto.store }
    private var sync: DittoSync { ditto.sync }

    private var subscription: DittoSyncSubscription?
    private var locationsObserver = AnyCancellable({})
    private var currentLocationDerivation = AnyCancellable({})

    init(dittoManager: DittoManager = .shared) {
        self.dittoManager = dittoManager
        startSubscription()
        observeAllLocations()
        deriveCurrentLocation()
        // Restore the persisted active location on launch — but only if it's
        // still one of the seven demo locations. Passes through the same setter
        // the UI uses, so the sync group is applied here too.
        if let saved = Settings.locationId {
            if LocationSeed.demoLocationIds.contains(saved) {
                setActiveLocation(saved)
            } else {
                // Stale id (e.g. a legacy custom location that no longer
                // exists). Clear it so the user is forced to re-pick, and reset
                // the sync group to default.
                setActiveLocation(nil)
            }
        }
    }

    /// Switch the active location. Persists, publishes, and applies Ditto's
    /// sync group. Pass `nil` to clear the selection and reset the sync group
    /// back to default.
    func setActiveLocation(_ locationId: String?) {
        Settings.locationId = locationId
        currentLocationId = locationId
        if let locationId {
            dittoManager.applySyncGroup(locationId: locationId)
        } else {
            dittoManager.resetSyncGroup()
        }
    }

    private func startSubscription() {
        do {
            subscription = try sync.registerSubscription(
                query: "SELECT * FROM \(Location.collectionName)"
            )
        } catch {
            reportSubscriptionFailure("subscribe locations", error)
        }
    }

    private func observeAllLocations() {
        locationsObserver = store.observePublisher(
            query: "SELECT * FROM \(Location.collectionName)",
            mapTo: Location.self
        )
        .map { locations in
            locations.filter { LocationSeed.demoLocationIds.contains($0.id) }
        }
        .replaceError(with: [])
        .assign(to: \.allLocations, on: self)
    }

    private func deriveCurrentLocation() {
        currentLocationDerivation = Publishers.CombineLatest($currentLocationId, $allLocations)
            .map { id, all in
                guard let id else { return nil }
                return all.first { $0.id == id }
            }
            .assign(to: \.currentLocation, on: self)
    }
}
